hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/InternalValueReader.java

```diff
@@ -0,0 +1,336 @@
+        validateType(position, FieldType.FLOAT_ARRAY);
+        in.position(position.getStreamPosition());
+        return in.readFloatArray();
+    }
+
+
+    private short[] readSingleShortArray(PortablePosition position) throws IOException {
+        validateType(position, FieldType.SHORT_ARRAY);
+        in.position(position.getStreamPosition());
+        return in.readShortArray();
+    }
+
+    private String[] readSingleUTFArray(PortablePosition position) throws IOException {
+        validateType(position, FieldType.UTF_ARRAY);
+        in.position(position.getStreamPosition());
+        return in.readUTFArray();
+    }
+
+    private Portable[] readSinglePortableArray(PortablePosition position) throws IOException {
+        in.position(position.getStreamPosition());
+        if (position.getLen() == Bits.NULL_ARRAY_LENGTH) {
+            return null;
+        }
+
+        validateType(position, FieldType.PORTABLE_ARRAY);
+        final Portable[] portables = new Portable[position.getLen()];
+        for (int index = 0; index < position.getLen(); index++) {
+            in.position(getPortableArrayCellPosition(in, position.getStreamPosition(), index));
+            portables[index] = serializer.readAndInitialize(in, position.getFactoryId(), position.getClassId());
+        }
+        return portables;
+    }
+
+
+    private <T> MultiResult<T> readMultiPosition(List<PortablePosition> positions) throws IOException {
+        MultiResult<T> result = new MultiResult<T>();
+        for (PortablePosition position : positions) {
+            if (!position.isNullOrEmpty()) {
+                T read = readSinglePosition(position);
+                result.add(read);
+            } else {
+                result.addNullOrEmptyTarget();
+            }
+        }
+        return result;
+    }
+
+    @SuppressWarnings("unchecked")
+    private <T> T readSinglePosition(PortablePosition position) throws IOException {
+        if (position.getIndex() >= 0) {
+            return readSinglePositionFromArray(position);
+        }
+        return readSinglePositionFromNonArray(position);
+    }
+
+    private PortablePosition findPositionForReading(String path) throws IOException {
+        try {
+            return PortablePositionNavigator.findPositionForReading(ctx, path, pathCursor);
+        } finally {
+            // The context is reset each time to enable its reuse in consecutive calls and avoid allocation
+            ctx.reset();
+        }
+    }
+
+    @SuppressWarnings({"checkstyle:cyclomaticcomplexity", "checkstyle:returncount", "unchecked"})
+    private <T> T readSinglePositionFromArray(PortablePosition position) throws IOException {
+        assert position.getType() != null : "Unsupported type read: null";
+        switch (position.getType()) {
+            case BYTE:
+            case BYTE_ARRAY:
+                return (T) Byte.valueOf(in.readByte(position.getStreamPosition()));
```

Comment:
when what is at the position is n array element and the element is an array, I was expecting to see a recursive kind of call here, is this correct code path? Is there a test where we have an array inside an array (do we even support it) ?

Reply:
I don't really know. This code was copy paste from the DefaultPortableReader. Not newly implemented. Let me check

Reply:
Here is how it works, when we get to this method the position is already resolved. We are already on a leaf that we can read directly. Anything recursive is handled in `findPositionForReading` method when we try to find the position to read.